backend/app/instance_provider.go

```diff
@@ -10,7 +11,7 @@ import (
 )
 
 // InstanceFactoryFunc factory method for creating app instances.
-type InstanceFactoryFunc func(settings backend.AppInstanceSettings) (instancemgmt.Instance, error)
+type InstanceFactoryFunc func(ctx context.Context, settings backend.AppInstanceSettings, cfg backend.Cfg) (instancemgmt.Instance, error)
```

Comment:
If we do this we'd need to do the following:
- Instead of altering this existing app and datasources instance managers (and underlying providers), we'd need to create new ones that would accept the new instance factory func and likely deprecate the old ones.
- Add new entry point signatures equivalent to `app.Manage` and `datasource.Manage` which would make use of the new instance manager, also making it play nice with the auto instance management.

The alternative is we break and people just add those two arguments to their existing factory funcs. 

Will wait to see what the rest of the team think!

Reply:
yep, that sucks but it's difficult to measure if that's better than breaking the API my gut feeling is "yes" if we consider the SDK as something stable (which maybe it isn't :P) 

wonder if using Go generics would help here :thinking:

Reply:
I've already suggested to introduce a breaking change. From the SDK readme
> This SDK is still in development. The protocol between the Grafana server and the plugin SDK is considered stable, but we might introduce breaking changes in the SDK...

Nit. One thing that feels odd/itchy is including the config as a third argument. It's perfectly fine, but wonder if this is the last time we need to include something that's not in the instancesetting struct. I'm thinking what about show it into `context.Context` and retrievable via `config.FromContext(ctx)`. Alternatively introduce another structure of the third argument that includes config as a field to allow it to be extended with additional fields if needed in the future.

Another alternative, introduce a `backend.Manage(...)` and `backend.InstanceFactoryFunc func(ctx context.Context, pluginContext backend.PluginContext` as the new default, handling both datasource and app. And then deprecate `app.Manage` and `datasource.Manage`. :shrug: :shrug: :shrug:

Reply:
Is having the config "absolutely" needed when instantiating the plugin? I know I suggested that plugins _may_ need to have access to feature flags (for example) when instantiating the data source but if that is not a requirement today we can get away without the breaking change if we just have the config available per request.

Reply:
> Nope not at all right now. We could always do a first step here by just adding the ability for this config to be accessed via PluginContext / context.Context, a worry about the rest later if necessary.

Sounds like a good path forward :+1:

Reply:
no ideal but in that case the client could be created during the first request once the required info is available

Reply:
>I guess it depends on if config is needed when creating HTTP clients/database connections etc?

Yep I spoke too soon. Was thinking about feature toggles which we don't necessarily _have_ to support, but for the likes of PDC which hook into the http client provider (which is used by many plugins at this stage as part of the instance factory)

>no ideal but in that case the client could be created during the first request once the required info is available

Yeah this kind of breaks our whole model so not really an option I don't think

Reply:
:cry: well, it seems I cannot protect people from having a breaking change :sweat_smile:. My +1 then to include the config to the context

Reply:
I agree with making this breaking change.

Like suggested by Marcus, I would also put `backend.Cfg` into `context.Context` (or add a new struct with `backend.Cfg` as a field, but `context.Context` works well in this case).